experimental/person_names_formatter/src/formatter.rs

```diff
@@ -0,0 +1,83 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use alloc::collections::btree_map::BTreeMap;
+
+use icu_locid::Locale;
+
+use crate::api::NameField;
+use crate::api::PersonName;
+use crate::api::PersonNamesFormatter;
+use crate::api::PersonNamesFormatterError;
+use crate::api::PreferredOrder;
+use crate::provider::PersonNamesFormattingDefinitionV1;
+
+///
+/// Default internal structure for person name.
+///
+pub struct DefaultPersonName {
+    person_data: BTreeMap<NameField, String>,
```

Comment:
_Especially_ since this is a short-lived object, it seems best to expand all the fields into struct fields instead of using a BTreeMap, which is expensive to create/destroy.

Reply:
How should we represent it in a Rust efficient way ? 

We can individually represent given name and surname in 2 separate data field, but since each data field can have any combination for modifier, each of them would end up having a map again.

```
pub given_name: BTreeMap<FieldModifier, String>
pub surname: BTreeMap<FieldModifier, String>
....
```

So, in an attempt to not use the BTreeMap, it would need to be flatten into... a vector of entries ?

```
pub struct FieldModifiedValue {
  pub all_caps:bool, 
  /// etc... bunch of other flags for all other modifiers.
  pub value: String
}
...
pub struct PersonName {
  ...
  pub given_names: Vec<FieldModifiedValue>
}
```

A person name structure can contains both {surname-core} {surname-prefix} at the same time.

Open to suggestion on how to best do it in Rust :)

Reply:
Moved to a LiteMap.

Done

Reply:
I think one way to make it work is have each field be a smallvec (or smallvec-based litemap) mapping modifiers to strings. So in the usual case you have just one.

Not needed to land this PR IMO